task-maker-async/src/store.rs

```diff
@@ -43,58 +82,591 @@ pub enum FileReadingOutcome {
+        }
+    }
+
+    /// Lists all the finalized computations that are similar to the given computation, i.e. for
+    /// which the first part of the hash matches.
+    pub fn similar_computations(&self, hash: ComputationHash) -> Vec<ComputationHash> {
+        let service = self.service.lock().unwrap();
+        let hash_top = hash.0;
+        match service.filesets.get(&hash_top) {
+            Some(variants) => variants
+                .iter()
+                .filter_map(|(k, v)| if v.finalized { Some(k) } else { None })
+                .map(|x| ComputationHash(hash_top.clone(), x.clone()))
+                .collect(),
+            None => vec![],
+        }
     }
 
-    /// Lists all the computations that are similar to the given computation, i.e. for which the
-    /// first part of the hash matches.
-    pub fn similar_computations(&self, _hash: ComputationHash) -> Vec<ComputationHash> {
-        todo!("");
+    /// Creates and starts a store, listening for RPCs on the given transport.
+    pub fn new_on_transport(
+        transport: impl Transport<Response<StoreResponse>, ClientMessage<StoreRequest>> + Send + 'static,
+    ) -> Self {
+        let token = CancellationToken::new();
+
+        let service = StoreService {
+            service: Arc::new(Mutex::new(StoreServiceImpl::new())),
+            cancellation_token: Arc::new(token),
+        };
+        let server = BaseChannel::with_defaults(transport);
+
+        {
+            let service = service.clone();
+            let token = service.cancellation_token.clone();
+            tokio::spawn(async move {
+                select! {
+                    _ = token.cancelled() => {}
+                    _ = server.execute(service.serve()) => {}
+                }
+            });
+        }
+
+        {
+            let token = service.cancellation_token.clone();
+            let service = service.clone();
+            tokio::spawn(async move {
+                let mut timer = interval(LEASE_LENGTH);
+                let timer = async {
+                    loop {
+                        let x = timer.tick().await;
+                        {
+                            service.service.lock().unwrap().clear_expired_leases();
+                        }
+                    }
+                };
+
+                select! {
+                    _ = token.cancelled() => {}
+                    _ = timer => {}
+                }
+            });
+        }
+        service
+    }
+
+    /// Stops the server.
+    pub fn stop(&self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+#[tarpc::server]
+impl Store for StoreService {
+    async fn create_or_open_input_file(
+        self,
+        context: Context,
+        hash: InputFileHash,
+    ) -> Result<FileSetHandle, Error> {
+        let mut service = self.service.lock().unwrap();
+        service.create_or_open_fileset(
+            ComputationHash(hash.0.clone(), hash.0),
```

Comment:
Can you explain why this is a `ComputationHash` even though this is an input file?